Client/game_sa/C2DEffectsSA.cpp

```diff
@@ -0,0 +1,1137 @@
+                    roadsign.flags = static_cast<std::uint8_t>(value);
+                    break;
+                case e2dEffectProperty::COLOR:
+                {
+                    SColor  color = TOCOLOR2SCOLOR(value);
+                    RwColor newColor = RwColor{color.R, color.G, color.B, color.A};
+
+                    pGame->GetModelInfo(model)->Store2DFXRoadsignColor(index, newColor);
+                    SetRoadsignTextColor(effect, newColor);
+                    break;
+                }
+                default:
+                    break;
+            }
+
+            break;
+        }
+        case e2dEffectType::ESCALATOR:
+        {
+            t2dEffectEscalator& escalator = effect->effect.escalator;
+            if (property == e2dEffectProperty::DIRECTION)
+                escalator.direction = static_cast<std::uint8_t>(value);
+
+            break;
+        }
+        default:
+            break;
+    }
+}
+
+void C2DEffectsSA::Set2DFXProperty(std::uint32_t model, std::uint32_t index, const e2dEffectProperty& property, bool value)
+{
+    C2DEffectSAInterface* effect = GetEffect(model, index);
+    if (!effect || effect->type != e2dEffectType::LIGHT || property != e2dEffectProperty::CORONA_REFLECTION)
+        return;
+
+    effect->effect.light.coronaEnableReflection = value;
+}
+
+void C2DEffectsSA::Set2DFXProperty(std::uint32_t model, std::uint32_t index, const e2dEffectProperty& property, std::string& value)
+{
+    C2DEffectSAInterface* effect = GetEffect(model, index);
+    if (!effect)
+        return;
+
+    switch (effect->type)
+    {
+        case e2dEffectType::LIGHT:
+        {
+            t2dEffectLight light = effect->effect.light;
+
+            switch (property)
+            {
+                case e2dEffectProperty::CORONA_NAME:
+                {
+                    if (light.coronaTex)
+                        RwTextureDestroy(light.coronaTex);
+
+                    CTxdStore_PushCurrentTxd();
+                    SetTextureDict(CTxdStore_FindTxdSlot("particle"));
+                    light.coronaTex = RwReadTexture(value.c_str(), nullptr);
+                    CTxdStore_PopCurrentTxd();
+                    break;
+                }
+                case e2dEffectProperty::SHADOW_NAME:
+                {
+                    if (light.shadowTex)
+                        RwTextureDestroy(light.shadowTex);
+
+                    CTxdStore_PushCurrentTxd();
+                    SetTextureDict(CTxdStore_FindTxdSlot("particle"));
+                    light.shadowTex = RwReadTexture(value.c_str(), nullptr);
+                    CTxdStore_PopCurrentTxd();
+                    break;
+                }
+                default:
+                    break;
+            }
+
+            break;
+        }
+        case e2dEffectType::PARTICLE:
+        {
+            if (property == e2dEffectProperty::PRT_NAME)
+                std::strncpy(effect->effect.particle.szName, value.c_str(), 24);
```

Comment:
this (and further strncpy calls below) do not result in a null-terminated string for the name if value is more than 24 characters. This will cause trouble later on.

Reply:
Would this solution be correct?

```cpp
effect->effect.particle.szName[23] = '\0';
```